File: projects/chiindii/iphone/Chindii/Chindii/FramedSerialProtocol.swift
//
//  FramedSerialProtocol.swift
//  Chindii
//
//  Created by Warren Janssens on 2016-01-18.
//  Copyright © 2016 Warren Janssens. All rights reserved.
//

import Foundation
import CoreBluetooth

protocol FramedSerialProtocolDelegate: NSObjectProtocol {
	func onMessage(message: FramedSerialMessage)
	func write(b: UInt8)
}

class FramedSerialProtocol : NSObject {
	//Error codes
	let NO_ERROR									= 0
	let INCOMING_ERROR_UNEXPECTED_START_OF_FRAME	= 1
	let INCOMING_ERROR_INVALID_CHECKSUM				= 2
	let INCOMING_ERROR_INVALID_LENGTH				= 3
	let INCOMING_ERROR_EXCEED_MAX_LENGTH			= 4
	let OUTGOING_ERROR_QUEUE_FULL					= 5
	
	//Special bytes
	let START : UInt8 = 0x7e
	let ESCAPE : UInt8 = 0x7d

	//Incoming state
	var position : Int = 0			// Current position in the frame
	var length : UInt8 = 0			// Frame length
	var command : UInt8 = 0			// Incoming message command
	var checksum : UInt8 = 0		// Checksum
	var escape : Bool = false 		// Escape byte seen, unescape next byte
	var error : Int = 0				// Error condition, ignore bytes until next frame start byte
	var data = [UInt8]()			// Incoming message
	var delegate : FramedSerialProtocolDelegate!
	
	//Convenience method to escape the given byte if needed
	func escapeByte(b : UInt8) {
		if (b == START || b == ESCAPE) {
			delegate.write(ESCAPE)
			delegate.write(b ^ 0x20);
		} else {
			delegate.write(b);
		}
	}
	
	func onMessage(message: [UInt8]) {
		for var i = 0; i < message.count; i++ {
			var b = message[i]

			if (error > 0) {
				if (b == START) {
					// recover from any previous error condition
					error = NO_ERROR;
					position = 0;
				}
				else {
					continue;
				}
			}
			
			if (position > 0 && b == START) {
				// unexpected start of frame
				error = INCOMING_ERROR_UNEXPECTED_START_OF_FRAME;
				continue;
			}
			if (position > 0 && b == ESCAPE) {
				// unescape next byte
				escape = true;
				continue;
			}
			if (escape) {
				// unescape current byte
				b = 0x20 ^ b;
				escape = false;
			}
			if (position > 1) { // start byte and length byte not included in checksum
				checksum += b;
			}
			
			switch(position) {
			case 0: // start frame
				position++;
				break;
			case 1: // length
				if (b == 0){
					error = INCOMING_ERROR_INVALID_LENGTH;
				}
				else {
					length = b;
					position++;
				}
				break;
			case 2:
				command = b;
				position++;
				break;
			default:
				if (position == Int(length + 2)) {
					if (checksum == 0xff) {
						delegate.onMessage(FramedSerialMessage(command : command, data : data))
						position = 0;
						checksum = 0;
					} else {
						error = INCOMING_ERROR_INVALID_CHECKSUM;
					}
					position = 0;
					checksum = 0;
				}
				else {
					data[position - 3] = b;
					position++;
				}
				break;
			}
		}
	}
	
	//Call this to write the entire message into the provided stream.
	func write(message : FramedSerialMessage) {
		let command = message.getCommand()
		let data = message.getData()
		let length = data.count;
		
		for (var position = 0; position <= (length + 3); position++) {
			switch(position){
			case 0:
				delegate.write(START);
				break;
			case 1:
				escapeByte(UInt8(length + 1));
				break;
			case 2:
				escapeByte(command);
				break;
			default:
				if (position - 3 == length){
					//Write checksum
					var result = command;
					
					for var i = 0; i < length; i++ {
						result = result &+ data[i]
					}
					
					escapeByte(0xff - result);
				}
				else {
					escapeByte(data[position - 3]);
				}
				break;
			}
		}
	}
	
	/*
	* Gets the latest error status code.  0 means no error, non-zero is error.
	*/
	func getError() -> Int { return error }
};

File: inc/common/FramedSerialProtocol/FramedSerialProtocol.cpp
#include "FramedSerialProtocol.h"

using namespace digitalcave;

FramedSerialMessage::FramedSerialMessage(uint8_t command, uint8_t maxSize){
	this->data = (uint8_t*) malloc(maxSize);
	this->command = command;
	this->maxSize = maxSize;
	length = 0;
}

FramedSerialMessage::~FramedSerialMessage(){
	free((void*) data);
}

FramedSerialMessage::FramedSerialMessage(uint8_t command, uint8_t* data, uint8_t length){
	this->data = (uint8_t*) malloc(maxSize);
	this->command = command;
	this->length = length;
	this->maxSize = length;
	for(uint8_t i = 0; i < length; i++){
		this->data[i] = data[i];
	}
}

void FramedSerialMessage::clone(FramedSerialMessage* m){
	command = m->command;
	for(uint8_t i = 0; i < maxSize; i++){
		data[i] = m->data[i];
	}
	length = m->length;
}

uint8_t FramedSerialMessage::getCommand(){
	return command;
}

uint8_t FramedSerialMessage::getLength(){
	return length;
}

uint8_t* FramedSerialMessage::getData(){
	return data;
}

void FramedSerialMessage::append(uint8_t b){
	data[length++] = b;
}

FramedSerialProtocol::FramedSerialProtocol(uint8_t maxSize){
	data = (uint8_t*) malloc(maxSize);
	this->maxSize = maxSize;
	position = 0;
	length = 0;
	command = 0;
	checksum = 0;
	escape = 0;
	error = 0;
}

FramedSerialProtocol::~FramedSerialProtocol(){
	free(data);
}

uint8_t FramedSerialProtocol::getError(){
	return error;
}

void FramedSerialProtocol::escapeByte(Stream* stream, uint8_t b){
	if (b == START || b == ESCAPE) {
		stream->write(ESCAPE);
		stream->write(b ^ 0x20);
	} else {
		stream->write(b);
	}
}

uint8_t FramedSerialProtocol::read(Stream* stream, FramedSerialMessage* result){
	uint8_t b;
	while (stream->read(&b)){
		if (position == 0 && b != START){
			//Garbage data, ignore
			continue;
		}
		
		if (error > 0){
			if (b == START) {
				// recover from any previous error condition
				error = NO_ERROR;
				position = 0;
			}
			else {
				continue;
			}
		}
	
		if (position > 0 && b == START) {
			// unexpected start of frame
			error = INCOMING_ERROR_UNEXPECTED_START_OF_FRAME;
			continue;
		}
		if (position > 0 && b == ESCAPE) {
			// unescape next byte
			escape = 1;
			continue;
		}
		if (escape) {
			// unescape current byte
			b = 0x20 ^ b;
			escape = 0;
		}
		if (position > 1) { // start byte and length byte not included in checksum
			checksum += b;
		}
	
		switch(position) {
			case 0: // start frame
				position++;
				break;
			case 1: // length
				if (b == 0){
					error = INCOMING_ERROR_INVALID_LENGTH;
				}
				else {
					length = b;
					position++;
				}
				break;
			case 2:
				command = b;
				position++;
				break;
			default:
				if ((position - 3) > maxSize){
					//Max size exceeded
					error = INCOMING_ERROR_EXCEED_MAX_LENGTH;
				}
				else if (position == (length + 2)) {
					if (checksum == 0xff) {
						FramedSerialMessage m(command, data, length - 1);
						result->clone(&m);
						position = 0;
						checksum = 0;
						return 1;
					} else {
						error = INCOMING_ERROR_INVALID_CHECKSUM;
					}
					position = 0;
					checksum = 0;
				}
				else {
					data[position - 3] = b;
					position++;
				}
				break;
		}
	}
	
	return 0;
}

void FramedSerialProtocol::write(Stream* stream, FramedSerialMessage* message){
	uint8_t length = message->getLength();
	uint8_t command = message->getCommand();
	uint8_t* data = message->getData();
	
	uint8_t checksum  = 0;
	for(uint8_t position = 0; position <= (length + 3); position++){
		switch(position){
			case 0:
				stream->write(START);
				break;
			case 1:
				escapeByte(stream, length + 1);
				break;
			case 2:
				escapeByte(stream, command);
				checksum = command;
				break;
			default:
				if (position - 3 == length){
					escapeByte(stream, 0xff - checksum);
				}
				else {
					uint8_t b = data[position - 3];
					escapeByte(stream, b);
					checksum += b;
				}
				break;
		}
	}
}

